### ISSUE TITLE
[C#] Generates invalid code when `groupSizeEncoding.numInGroup` is a `uint32`

### ISSUE THREAD

**Issue by ZachBray** [real-logic/simple-binary-encoding#567]
We're using SBE as the basis for a snapshot serialization tool. We're running into an issue where a team that uses our tool needs to store a list with `> 2^16` elements.

### `uint32`

If I specify the group size encoding should use a `uint32` the C# code generator generates invalid code.

```xml
        <composite name="groupSizeEncodingU32" description="Repeating group dimensions">
            <type name="blockLength" primitiveType="uint16"/>
            <type name="numInGroup" primitiveType="uint32"/>
        </composite>
```

```csharp
        public sealed partial class MyListGroup
        {
            private readonly GroupSizeEncodingU32 _dimensions = new GroupSizeEncodingU32();
            // ...
            private int _count;
            // ...

            public void WrapForDecode(MessageWithList32 parentMessage, DirectBuffer buffer, int actingVersion)
            {
                // ...
                _count = _dimensions.NumInGroup;
                //             ^^^^^ error here because RHS (uint) may not fit in LHS (int)
                // ...
            }
```

### `int32`

If I specify the group size encoding should use an `int32` the schema parser complains. This seems fair, as a negative length list wouldn't make much sense.

```xml
        <composite name="groupSizeEncoding" description="Repeating group dimensions">
            <type name="blockLength" primitiveType="uint16"/>
            <type name="numInGroup" primitiveType="int32"/>
        </composite>
```

```bash
ERROR: at <ns2:message name="messageWithList32"> <group name="myList"> "numInGroup" must be unsigned type
```

### What is the best way to remedy this issue?

_My preferred option is currently **C**._

#### A. Fix the C# code generation to support `uint32`

This has the downside that the visible API would change. It is a breaking change for existing consumers of the library.

#### B. Relax the must-be-unsigned constraint

This would make the schemas look less sane; however, AFAIK there are no unsigned numbers in Java and .NET uses `int` rather than `uint` to index arrays etc. (as VB doesn't support unsigned numbers either) so there is precedent.

#### C. Apply a `maxValue` constraint to make the `uint32` a `uint31` and cast it to an `int32`

This change is the smallest. We could fail-fast in the C# code generation if the `maxValue` (which we can provide in the schema) is not representable by an `int` making it a somewhat "safe" cast.

```xml
        <composite name="groupSizeEncodingU32" description="Repeating group dimensions">
            <type name="blockLength" primitiveType="uint16"/>
            <type name="numInGroup" primitiveType="uint32" maxValue="2147483647"  />
        </composite>
```

#### D. Something I haven't thought of...

**Comment by mjpt777** [comment]
I'd go with an option D. That is don't put so many elements in a repeating group. Consider that they will be a single message that is huge. This will have to be assembled into a single message by Aeron or any other system. All the memory needs to be resident at the same time for a single huge message.

Decompose your elements of the repeating group into their own messages to be better suited to streaming.

As for the actual bug then fix the code to generate the right type with apply limits.

**Comment by ZachBray** [comment]
> All the memory needs to be resident at the same time for a single huge message

The repeated group in question comprises references encoded as int identifiers. The team say they have a case where ~ 80,000 references are needed. I think this will make the message ~ 350KB. Other than, smaller-the-better do you have any rules of thumb here?

> Decompose your elements of the repeating group into their own messages to be better suited to streaming

Ideally, we'd be able to "reverse" the direction of reference for serialization a la relational DBs. For example, if we have a relationship `A 1--* B` we'd encode a single reference to an `A` in each serialized representation of `B`. However, as our snapshot process is serializing an object graph [0] we have far more "interesting" relationships than this simple case.

Perhaps we could have encoded "objA.field references objB" as its own message; however, this would add a lot of complexity. It would need to incorporate addressing of fields for a start. I don't think this would have been realistic in the short timeframe we had.

> fix the code to generate the right type with apply limits

Just to be clear here, is this A or C or something else? Both A and C involve changes to the code generation. I think this is C though.

[0] - Our solution needed to be retrofitted to a relatively large existing model so we didn't have the option of avoiding serializing a graph.

**Comment by mjpt777** [comment]
Option B is not supported by the specification. I would like to support signed values and even use the likes to -1 to signal null or not present, however the spec will not allow this either.

C is probably your best option for now. I've changed the Java so it could read these messages if necessary but the compiler will warn as it is not really supported or prohibited under the spec.

Messages sizes more than a few megabytes is bad design. Smaller the better. If you don't think about this in the design they it will always come back to bite in some way. Going with C only kicks the can down the road a little. It is usually best to keep entity relationships separate from the entities to allow for more flexibility but this is a much larger topic.

**Comment by ZachBray** [comment]
Thank you. I'll raise a PR for the C# side.

> Going with C only kicks the can down the road a little. It is usually best to keep entity relationships separate from the entities to allow for more flexibility but this is a much larger topic.

Will bear this in mind and see if we can move towards a better design soon.

**Comment by ZachBray** [comment]
Should be fixed in #568
